src/sync.rs

```diff
@@ -1290,6 +1600,56 @@ mod tests {
         });
     }
 
+    #[cfg(all(Py_3_14, not(Py_LIMITED_API)))]
+    #[cfg(not(target_arch = "wasm32"))] // We are building wasm Python with pthreads disabled
+    #[test]
+    fn test_critical_section_mutex2_two_containers() {
+        let (m1, m2) = (PyMutex::new(vec![1, 2, 3]), PyMutex::new(vec![4, 5]));
+
+        std::thread::scope(|s| {
+            s.spawn(|| {
+                Python::attach(|py| {
+                    with_critical_section_mutex2(py, &m1, &m2, |mut v1, v2| {
+                        // v1.extend(v1)
+                        // SAFETY: we never call back into the python interpreter inside this critical section
+                        let vec1 = unsafe { v1.get_mut() };
+                        let vec2 = unsafe { v2.get() };
+                        vec1.extend(vec2.iter());
+                    })
+                });
+            });
+            s.spawn(|| {
+                Python::attach(|py| {
+                    with_critical_section_mutex2(py, &m1, &m2, |v1, mut v2| {
+                        // v2.extend(v1)
+                        // SAFETY: we never call back into the python interpreter inside this critical section
+                        let vec1 = unsafe { v1.get() };
+                        let vec2 = unsafe { v2.get_mut() };
+                        vec2.extend(vec1.iter());
+                    })
+                });
+            });
+        });
+
+        // execution order is not guaranteed, so we need to check both
+        // NB: extend should be atomic, items must not be interleaved
+        // v1.extend(v2)
+        // v2.extend(v1)
+        let expected1_vec1 = vec![1, 2, 3, 4, 5];
+        let expected1_vec2 = vec![4, 5, 1, 2, 3, 4, 5];
+        // v2.extend(v1)
+        // v1.extend(v2)
+        let expected2_vec1 = vec![1, 2, 3, 4, 5, 1, 2, 3];
+        let expected2_vec2 = vec![4, 5, 1, 2, 3];
+
+        let v1 = m1.lock().unwrap();
+        let v2 = m2.lock().unwrap();
+        assert!(
+            ((*v1).eq(&expected1_vec1) && (*v2).eq(&expected1_vec2))
+                || ((*v1).eq(&expected2_vec1) && (*v2).eq(&expected2_vec2))
```

Comment:
Condition might read a bit easier if we deref on the assignments (makes the locks last to the end of scope but not a problem here).


```suggestion
        let v1 = &*m1.lock().unwrap();
        let v2 = &*m2.lock().unwrap();
        assert!(
            (v1.eq(&expected1_vec1) && v2.eq(&expected1_vec2))
                || (v1.eq(&expected2_vec1) && v2.eq(&expected2_vec2))
```